### PR TITLE
DAOS-6367 object: workaround for OC_SX/OC_RP_XSF

### PR DESCRIPTION
**Pull request by gnailzenh**
DAOS-6367 object: workaround for OC_SX/OC_RP_XSF
    
The main issue of OC_SX/OC_RP_XSF objects is these objects
cannot be rebuilt or rebalanced. This is the nature of their
definition: spreading to ALL targets in the pool. "ALL" can
be changed over time when user including/excluding targets
from the pool, so there is no way for the placement algorithm
to generate the same layout after membership change of a pool.
    
This patch implements a function:
      daos_oclass_fit_max(int oc_id, int domain_nr,
                          int target_nr, int *oc_id_p)
    
If the input @oc_id is OC_SX or any class which requires more
targets than the pool actually has (target_nr), or the input
@oc_id is OC_XSF or any class which requires more domains
than the pool actually has (domain_nr), this function can map
the input class to another class which can fit into the pool.
    
It also changes DFS to call this function, so it can convert
OC_SX/OC_RP_XSF to anthoer class with exact targets number.
    
This patch also defines more object classes.
    
TODO:
- add more object classes to the class table
- move the DFS object class select function to a generic module

Signed-off-by: Liang Zhen <liang.zhen@intel.com>

### CODE REVIEW

**Review by mchaarawi — DISMISSED** [review]
this looks OK to me.
part of the to-do should be to have something public to use from HDF5, and other apps / middleware that can be built on top of this.

**Review by byron-marohn — DISMISSED** [review]
Maybe we could put a warning in someplace for people who still use the XSF or GX object classes? Maybe in the OID generation function? "Warning this object is widely striped and will likely be damaged by changing the DAOS server configuration (adding servers, etc.)" ?

**Comment by gnailzenh** [comment]
sorry, need to fix an issue, will refresh soon

**Comment by gnailzenh** [comment]
There are main changes:
- instead of storing object class in property, now we store SB oid and root oid. (the property is called DAOS_PROP_CO_ROOTS, I discussed with Johann and he thinks it could be used by other middlewares) 
- the patch gets target number and domain number from placement map now, instead of pool. This is because we do need domain number for OC_RP_XSF, pool client does not have this information.

File: src/object/obj_class.c [hunk 1]
```diff
@@ -58,7 +68,176 @@ static struct daos_obj_class daos_obj_classes[] = {
...
+		.oc_id		= OC_S8K,
+		{
+			.ca_schema		= DAOS_OS_STRIPED,
+			.ca_resil		= DAOS_RES_REPL,
+			.ca_grp_nr		= (8 << 10),
+			.ca_rp_nr		= 1,
+		},
+	},
 	{
 		.oc_name	= "SX",
```
Comment (wangdi1) [inline src/object/obj_class.c:242]: Should we still keep SX for now?

File: src/client/dfs/dfs.c [hunk 2]
```diff
@@ -1270,16 +1324,20 @@ dfs_mount(daos_handle_t poh, daos_handle_t coh, int flags, dfs_t **_dfs)
 		D_GOTO(err_dfs, rc = daos_der2errno(rc));
 	}
 
+	entry = daos_prop_entry_get(prop, DAOS_PROP_CO_ROOTS);
+	D_ASSERT(entry != NULL);
+	roots = (struct daos_prop_co_roots *)entry->dpe_val_ptr;
+	dfs->sb_oid = roots->cr_oids[0];
+	dfs->root.oid = roots->cr_oids[1];
+	dfs->root.parent_oid = dfs->sb_oid;
```
Comment (wangdi1) [inline src/client/dfs/dfs.c:1332]: Do we need valid the the root id? in case it open an old version container?

File: src/common/proc.c [hunk 3]
```diff
@@ -86,23 +87,38 @@ crt_proc_prop_entries(crt_proc_t proc, daos_prop_t *prop)
...
+
+		} else if (entry->dpe_type == DAOS_PROP_PO_SVC_LIST) {
 			rc = crt_proc_d_rank_list_t(proc,
 					(d_rank_list_t **)&entry->dpe_val_ptr);
-		else
+		} else if (entry->dpe_type == DAOS_PROP_CO_ROOTS) {
+			struct daos_prop_co_roots *roots;
+
+			if (opc == CRT_PROC_DECODE)
+				D_ALLOC(entry->dpe_val_ptr, sizeof(*roots));
```
Comment (wangdi1) [inline src/common/proc.c:112]: check allocation result?

File: src/include/daos_prop.h [hunk 4]
```diff
@@ -183,9 +183,16 @@ enum daos_cont_props {
...
 	 */
 	DAOS_PROP_CO_DEDUP_THRESHOLD,
+	/** First citizon objects of container, see \a daos_cont_root_oids */
+	DAOS_PROP_CO_ROOTS,
 	DAOS_PROP_CO_MAX,
 };
 
+/* first citizen objects of a container, stored as container property */
+struct daos_prop_co_roots {
+	daos_obj_id_t	cr_oids[4];
```
Comment (wangdi1) [inline src/include/daos_prop.h:193]: Just curious why reserve 4 oids here?  probably add some version here in case you want add more here?

**Review by wangdi1 — DISMISSED** [review]

File: src/include/daos_prop.h (hunk 4 above)
Comment (gnailzenh) [reply src/include/daos_prop.h:193]: Yes, just reserved 2 more in case other middlewares need more

File: src/client/dfs/dfs.c (hunk 2 above)
Comment (gnailzenh) [reply src/client/dfs/dfs.c:1334]: I can add a version number, but I think this is responsibility of middleware, not DAOS

File: src/object/obj_class.c (hunk 1 above)
Comment (gnailzenh) [reply src/object/obj_class.c:242]: Yes, we should, user uses it to indicate it is a "large" object.

**Comment by gnailzenh** [comment]
@mchaarawi one of those failures is HDF5 VOL test, I think it's an issue of the test itself because it's assuming class ID equals to a number value. I agree after 1.2 we can assume that, but not now, I'd suggest to disable those checks.

The second failure is OSA, I will work in Byron on it, but this patch cannot wait for too long.

**Comment by byron-marohn** [comment]
@rpadma2 does the OSA drain failure here look familiar?

**Comment by rpadma2** [comment]
@byron-marohn : When did you merge with master? Looks like daos_io_server crashed.

2021-02-01 15:05:02,139 process          L0479 DEBUG| [stdout] [1,0]<stdout>:DEBUG 15:05:02.139272 instance_drpc.go:77: dRPC to index 0 (rank 0): PoolQuery
2021-02-01 15:05:02,151 process          L0479 DEBUG| [stderr] [1,2]<stderr>:ERROR: daos_io_server:0 *** Error in `/usr/bin/daos_io_server': realloc(): invalid next size: 0x00007fd3e5a25d00 ***
2021-02-01 15:05:02,152 process          L0479 DEBUG| [stderr] [1,2]<stderr>:ERROR: daos_io_server:0 ======= Backtrace: =========
2021-02-01 15:05:02,152 process          L0479 DEBUG| [stderr] [1,2]<stderr>:ERROR: daos_io_server:0 /lib64/libc.so.6(+0x7f3e4)[0x7fd418dcb3e4]
2021-02-01 15:05:02,152 process          L0479 DEBUG| [stderr] [1,2]<stderr>:ERROR: daos_io_server:0 /lib64/libc.so.6(+0x847d1)[0x7fd418dd07d1]
2021-02-01 15:05:02,152 process          L0479 DEBUG| [stderr] [1,2]<stderr>:ERROR: daos_io_server:0 /lib64/libc.so.6(realloc+0x1d2)[0x7fd418dd1d82]
2021-02-01 15:05:02,152 process          L0479 DEBUG| [stderr] [1,2]<stderr>:/lib64/libdaos.so.1(+0xc37b9)[0x7fd41a2e87b9]
2021-02-01 15:05:02,152 process          L0479 DEBUG| [stderr] [1,2]<stderr>:/lib64/libdaos.so.1(+0xc59b4)[0x7fd41a2ea9b4]
2021-02-01 15:05:02,152 process          L0479 DEBUG| [stderr] [1,2]<stderr>:ERROR: daos_io_server:0 /lib64/libdaos.so.1(obj_enum_iterate+0x1d4)[0x7fd41a2ed064]
2021-02-01 15:05:02,152 process          L0479 DEBUG| [stderr] [1,2]<stderr>:/lib64/libdaos.so.1(dss_enum_unpack+0x32d)[0x7fd41a2ed86d]
2021-02-01 15:05:02,153 process          L0479 DEBUG| [stderr] [1,2]<stderr>:/usr/lib64/daos_srv/libobj.so(+0x6995b)[0x7fd4106e795b]
2021-02-01 15:05:02,153 process          L0479 DEBUG| [stderr] [1,2]<stderr>:/usr/lib64/daos_srv/libobj.so(+0x6ad33)[0x7fd4106e8d33]
2021-02-01 15:05:02,153 process          L0479 DEBUG| [stderr] [1,2]<stderr>:/lib64/libabt.so.0(+0x10d8f)[0x7fd4197b5d8f]
2021-02-01 15:05:02,153 process          L0479 DEBUG| [stderr] [1,2]<stderr>:ERROR: daos_io_server:0 /lib64/libabt.so.0(+0x11501)[0x7fd4197b6501]

**Comment by mchaarawi** [comment]
> @mchaarawi one of those failures is HDF5 VOL test, I think it's an issue of the test itself because it's assuming class ID equals to a number value. I agree after 1.2 we can assume that, but not now, I'd suggest to disable those checks.
> 

I don't understand this failure. i thought your patch only changes the DFS case? the HDF5 VOL does not use DFS.
if someone else is using SX though the oid API that we have not changed, why does it get back a different oclass with a number value?

**Comment by mchaarawi** [comment]
> > @mchaarawi one of those failures is HDF5 VOL test, I think it's an issue of the test itself because it's assuming class ID equals to a number value. I agree after 1.2 we can assume that, but not now, I'd suggest to disable those checks.
> 
> I don't understand this failure. i thought your patch only changes the DFS case? the HDF5 VOL does not use DFS.
> if someone else is using SX though the oid API that we have not changed, why does it get back a different oclass with a number value?

actually i understand.. the problem is that we changed the value in the oclass enum for SX, and we would need to retrigger the HDF5 vol build to generate new RPMs, since this is an API change.
the enum value of SX in current master is actually for S256 in your PR. so the current vol RPM is incompatible with your PR.

**Pull request merged**